MAINT: Add better error handling in linalg.norm for vectors and clarify it in documentation

Jimver opened the pull request:
From PR #15535 the conclusion was that the documentation for linalg.norm needed to be improved and a better error message needed to be shown when calling `norm` with a vector and `ord='fro'`. This PR enhances the documentation for the linalg.norm() function to explicitly state that the frobenius norm is only defined for two dimensional arrays. See issue #15533 and PR #15535.

Jimver committed:
DOC: clarify fro usage in linalg.norm documentation (see #15533)

Jimver commented:
I'd also like to implement a better error message, should I do this in a separate one or this one?

rossbar commented:
Thanks for your interest @Jimver , this PR should definitely include improved error-handling.

Jimver committed:
ENH: Add improved error handling when getting frobenius norm from a vector (see #15533)

Jimver committed:
TST: Fix comment in linalg norm test

Jimver committed:
MAINT: Change exception to ValueError in linalg norm

Jimver commented:
> 
> 
> Thanks for your interest @Jimver , this PR should definitely include improved error-handling.

@rossbar I have added improved error handling to this PR, I also fixed a comment in a test that actually tests for this specific case of calculating the frobenius norm of a vector.

Please let me know if this error handling is good or if I should change the error message.

rossbar reviewed (changes requested):
A few initial comments:
 - The explicit case for `ord in ('f', 'fro')` is on the right track, but we should take this opportunity to fixup the rest of the code there. I'm confused about the try-except statement (the `ord + 1` doesn't appear to do anything meaningful except as a *very* roundabout type test?)
 - Note that there is another special kwarg "nuc" that should get the same treatment (currently also seems to be caught by the `try-except`)
 - Make sure any proposed changes obey the 80 char line limit.

I can take a closer look after the next round of changes.

Jimver committed:
ENH: Add more check statements in linalg norm for vectors

Jimver commented:
@rossbar I processed the feedback in the latest commit, I use an `isinstance` check if ord is an integer. I also added a separate check for `nuc`.

rossbar commented on numpy/linalg/linalg.py:
```diff
@@ -2556,11 +2561,15 @@ def norm(x, ord=None, axis=None, keepdims=False):
+                             "vectors only support integer norm orders. For "
+                             "example for the L2 norm use ord=2 instead")
+        elif ord is 'nuc':
+            raise ValueError("Nuclear norm is not supported for vectors")
```
Since none of the named kwargs ('fro', 'nuc') are valid for vectors, it seems to me these checks could be consolidated.
```suggestion
        elif ord in ('f', 'fro', 'nuc'):
            raise ValueError(f"Invalid norm order '{ord}' for vectors")
```

rossbar commented on numpy/linalg/linalg.py:
```diff
@@ -2556,11 +2561,15 @@ def norm(x, ord=None, axis=None, keepdims=False):
+        elif ord is 'nuc':
+            raise ValueError("Nuclear norm is not supported for vectors")
+        elif not isinstance(ord, numbers.Integral):
+            raise ValueError("Invalid norm order for vectors.")
```
Is this additional condition necessary? Are there cases outside of the 'nuc' and 'fro' for which the `try-catch` you removed are on the path? Other (invalid) str kwargs?

If you can verify that doing so is sufficient for replacing the original `try-except` block that was removed, you could consider further consolidating the above check to:

```python
elif isinstance(ord, str):
    raise ValueError(f"Invalid norm order '{ord}' for vectors")
```

Either way, I think a check based on `numbers.Integral` should be avoided.

Jimver replied:
Yeah I agree it is not optimal. I changed it to catch any string as you mentioned in case people have a typo in their `ord` string.

rossbar commented on numpy/linalg/linalg.py:
```diff
@@ -2431,6 +2432,10 @@ def norm(x, ord=None, axis=None, keepdims=False):
     The Frobenius norm is given by [1]_:
 
         :math:`||A||_F = [\\sum_{i,j} abs(a_{i,j})^2]^{1/2}`
+        
```
Personally, I don't know whether this extra couple sentences is necessary, but that's an entirely subjective take. If anything, I would try to shorten the additional info while being clear and explicit, maybe something like:

```
Both the Frobenius and nuclear norm orders are only defined for matrices
and raise a ValueError when `x.ndim != 2`.
```

Jimver replied:
IMO it is necessary as some people think the Frobenius norm works for vectors according to [Wolfram](https://mathworld.wolfram.com/FrobeniusNorm.html) so this extra clarification makes this more explicitly clear. I changed it to your shorter description to make it more succinct.

rossbar replied:
Makes sense to me. One last nitpick - could it be moved so that the line about the nuclear norm is above the new sentence? I think it flows better if it's:
 - explain Frobenius norm
 - explain nuclear norm
 - explain special cases about both fro/nuc for vectors

Jimver committed:
ENH: Consolidate linalg norm errors into single statement

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

Jimver committed:
ENH: Change linalg norm exception to catch any string

Jimver committed:
TST: Add a test case for arbitrary ord string in linarg norm

Jimver committed:
DOC: Shorten comment about frobenius norm in linalg.norm()

rossbar commented on numpy/linalg/linalg.py:
```diff
@@ -15,6 +15,7 @@
            'LinAlgError', 'multi_dot']
 
 import functools
+import numbers
```
```suggestion
```
No longer used

rossbar commented on numpy/linalg/linalg.py:
```diff
@@ -2556,11 +2560,9 @@ def norm(x, ord=None, axis=None, keepdims=False):
             s = (x.conj() * x).real
             return sqrt(add.reduce(s, axis=axis, keepdims=keepdims))
+        elif isinstance(ord, str):
+            raise ValueError(f"Invalid norm order '{ord}' for vectors")
```
```suggestion
            return sqrt(add.reduce(s, axis=axis, keepdims=keepdims))
        # None of the str-type keywords for ord ('fro', 'nuc') 
        # are valid for vectors
        elif isinstance(ord, str):
            raise ValueError(f"Invalid norm order '{ord}' for vectors")
```
Since it is not immediately obvious what this branch does, a comment would likely be useful for future readers of the code

rossbar commented on numpy/linalg/linalg.py:
```diff
@@ -2431,6 +2432,9 @@ def norm(x, ord=None, axis=None, keepdims=False):
         :math:`||A||_F = [\\sum_{i,j} abs(a_{i,j})^2]^{1/2}`
+        
+    Both the Frobenius and nuclear norm orders are only defined for
+    matrices and raise a ValueError when `x.ndim != 2`.
```
```suggestion
    matrices and raise a ValueError when ``x.ndim != 2``.
```
My fault - I originally suggested markdown syntax, not rST

Jimver replied:
Yes I agree, fixed it just now!

rossbar reviewed (changes requested):
Thanks for the changes - I think this is just about there, one last round of minor tweaks. After that we can ping another reviewer for another set of eyes.

Thanks for adding the additional test as well - that's definitely an improvement from a clarity & coverage standpoint :+1:

Jimver committed:
MAINT: Remove unused import in linalg

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

Jimver committed:
DOC: Add comment to linalg edge case

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

Jimver committed:
DOC: restructure linalg.norm documentation

Jimver commented:
@rossbar I processed your feedback, could you ping other reviewers? I am not so familiar with who I can ping.

WarrenWeckesser commented:
I'm taking a look now.

eric-wieser reviewed (approved):
Looks good, although PR title should probably be updated to reflect the final patch (which a maintainer can then use in the squashed commit message.

WarrenWeckesser reviewed (approved):
LGTM.

WarrenWeckesser commented:
The test failures are not related to this PR (something is going wrong with PyPy testing).

Jimver commented:
> The test failures are not related to this PR (something is going wrong with PyPy testing).

Yes when looking at the logs it is something to do with ubuntu failing to downloading dependencies

WarrenWeckesser commented:
Do we consider this a fix for gh-15533, so it can be closed when this is merged?

WarrenWeckesser commented:
Merged.  Thanks @Jimver!  And thanks @rossbar and @eric-wieser for reviews.
